integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/reference/FieldReferenceIT.java

```diff
@@ -0,0 +1,524 @@
+				.where( f -> f.matchAll() )
+		).hasHitsAnyOrder(
+				"a_0", "a_1", "a_2"
+		);
+	}
+
+	private static class IndexedEntity {
+		private int primitiveInteger;
+		private Double wrapperDouble;
+		private String string;
+		private String stringProjectedAsBytes;
+		private LocalDate localDate;
+		private float[] vector;
+
+		private List<String> stringList;
+
+		private EmbeddedEntity nestedEmbeddedEntity;
+		private EmbeddedEntity flattenedEmbeddedEntity;
+
+		private List<EmbeddedEntity> embeddedEntityList;
+
+		// @IndexedEmbedded(includeDepth = 2,
+		// 		structure = ObjectStructure.NESTED,
+		// 		excludePaths = {
+		// 				"parent.parent.primitiveInteger"
+		// 		// ... (other paths so that only string and vector fields are included)
+		// 		}
+		// )
+		private IndexedEntity parent;
+
+	}
+
+	private static class EmbeddedEntity {
+		private String string;
+		private SomeEnum someEnum;
+	}
+
+	private enum SomeEnum {
+		VALUE_1, VALUE_2, VALUE_3
+	}
+
+	public static class IndexedEntity_ {
+		public static ValueFieldReference<Integer, Integer, Integer> primitiveInteger;
+		public static ValueFieldReference<Double, Double, Double> wrapperDouble;
+		public static ValueFieldReference<String, String, String> string;
+		public static ValueFieldReference<String, String, byte[]> stringProjectedAsBytes;
+		public static ValueFieldReference<LocalDate, LocalDate, LocalDate> localDate;
+		public static ValueFieldReference<float[], float[], float[]> vector;
+
+		public static ValueFieldReference<String, String, String> stringList;
+		public static NestedEmbeddedEntity_ nestedEmbeddedEntity;
+		public static EmbeddedEntityList_ embeddedEntityList;
+
+		public static FlattenedEmbeddedEntity_ flattenedEmbeddedEntity;
+
+		public static Parent_ parent;
+
+		static {
+			primitiveInteger = ValueFieldReference.of( "primitiveInteger", Integer.class, Integer.class, Integer.class );
+			wrapperDouble = ValueFieldReference.of( "wrapperDouble", Double.class, Double.class, Double.class );
+			string = ValueFieldReference.of( "string", String.class, String.class, String.class );
+			stringProjectedAsBytes =
+					ValueFieldReference.of( "stringProjectedAsBytes", String.class, String.class, byte[].class );
+			localDate = ValueFieldReference.of( "localDate", LocalDate.class, LocalDate.class, LocalDate.class );
+			vector = ValueFieldReference.of( "vector", float[].class, float[].class, float[].class );
+			stringList = ValueFieldReference.of( "stringList", String.class, String.class, String.class );
+
+			flattenedEmbeddedEntity = new FlattenedEmbeddedEntity_( "flattenedEmbeddedEntity" );
+			nestedEmbeddedEntity = new NestedEmbeddedEntity_( "nestedEmbeddedEntity" );
+			embeddedEntityList = new EmbeddedEntityList_( "embeddedEntityList" );
+			parent = new Parent_( "parent" );
+		}
+
+		public static class NestedEmbeddedEntity_ extends ObjectFieldReferenceImpl implements NestedObjectFieldReference {
```

Comment:
I went with an approach to make all these object fields classes as nested classes.
I was considering making them all at the same level, but then the class names will grow the deeper the fields are e.g. `IndexedEntity_Parent_Parent_NestedEmbeddedEntity_`